crates/polars-parquet/src/parquet/encoding/hybrid_rle/encoder.rs

```diff
@@ -3,26 +3,118 @@ use std::io::Write;
 use super::bitpacked_encode;
 use crate::parquet::encoding::{bitpacked, ceil8, uleb128};
 
-/// RLE-hybrid encoding of `u32`. This currently only yields bitpacked values.
+// Arbitrary value that balances memory usage and storage overhead
+const MAX_VALUES_PER_LITERAL_RUN: usize = (1 << 10) * 8;
+
+// Iterator over an array up to a specified final index
+struct ArrayIterator<'a, T> {
+    array: &'a [T],
+    index: usize,
+    final_idx: usize,
+}
+
+impl<'a, T: Copy> ArrayIterator<'a, T> {
+    fn new(array: &'a [T], final_idx: usize) -> Self {
+        ArrayIterator {
+            array,
+            index: 0,
+            final_idx,
+        }
+    }
+}
+
+impl<'a, T: Copy> Iterator for ArrayIterator<'a, T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.final_idx {
+            self.index += 1;
+            Some(self.array[self.index - 1])
+        } else {
+            None
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = self.final_idx - self.index;
+        (remaining, Some(remaining))
+    }
+}
+
+#[allow(clippy::comparison_chain)]
 pub fn encode_u32<W: Write, I: Iterator<Item = u32>>(
     writer: &mut W,
     iterator: I,
     num_bits: u32,
 ) -> std::io::Result<()> {
-    let num_bits = num_bits as u8;
-    // the length of the iterator.
-    let length = iterator.size_hint().1.unwrap();
-
-    // write the length + indicator
-    let mut header = ceil8(length) as u64;
-    header <<= 1;
-    header |= 1; // it is bitpacked => first bit is set
-    let mut container = [0; 10];
-    let used = uleb128::encode(header, &mut container);
-    writer.write_all(&container[..used])?;
-
-    bitpacked_encode_u32(writer, iterator, num_bits as usize)?;
-
+    let mut consecutive_repeats: usize = 0;
+    let mut buffered_bits = [0; MAX_VALUES_PER_LITERAL_RUN];
+    let mut buffer_idx = 0;
+    let mut literal_run_idx = 0;
+    let mut previous_val = 0;
+    for val in iterator {
+        if val == previous_val {
+            consecutive_repeats += 1;
+            // Run is long enough to RLE, no need to buffer values
+            if consecutive_repeats > 8 {
+                continue;
+            // Ensure literal run has multiple of 8 values
+            // Take from consecutive repeats if needed to pad up
+            } else if consecutive_repeats == 8 {
```

Comment:
Can we move this branch maybe under a `if consecutive_repeats >= 8`, I think that would save the extra branch 1/8 consec values.